### ISSUE TITLE
Defined Extension Sets Should Be Immutable

### ISSUE THREAD

**Issue by mjordan56** [dart-lang/markdown#303]
The pre-defined extension sets in extension_set.dart, ExtensionSet.none, ExtensionSet.commonMark, ExtensionSet.gitHubWeb, and ExtensionSet.gitHubFlavored, should be immutable. The current implementation is mutable and can be easily changed by the consumer of the markdown package.

An example of why this is a problem, in debugging the flutter_markdown package there was code that added a custom inline syntax object to the ExtensionSet.gitHubFlavored extension set. Each time the Flutter Markdown widget was updated, a custom inline syntax object was created and added to the inlineSyntaxes list of the ExtensionSet.gitHubFlavored. Since ExtensionSet.gitHubFlavored is a static property the instances of the custom inline objects increased by one with every call.

The simplest solution to address this problem is to make the blockSyntaxes and inlineSyntaxes lists of the four extension set definitions unmodifiable like this:

```
import 'block_parser.dart';
import 'inline_parser.dart';

/// ExtensionSets provide a simple grouping mechanism for common Markdown
/// flavors.
///
/// For example, the [gitHubFlavored] set of syntax extensions allows users to
/// output HTML from their Markdown in a similar fashion to GitHub's parsing.
class ExtensionSet {
  /// The [ExtensionSet.none] extension set renders Markdown similar to
  /// [Markdown.pl].
  ///
  /// However, this set does not render _exactly_ the same as Markdown.pl;
  /// rather it is more-or-less the CommonMark standard of Markdown, without
  /// fenced code blocks, or inline HTML.
  ///
  /// [Markdown.pl]: http://daringfireball.net/projects/markdown/syntax
  static final ExtensionSet none = ExtensionSet(
    List<BlockSyntax>.unmodifiable(<BlockSyntax>[]),
    List<InlineSyntax>.unmodifiable(<InlineSyntax>[]),
  );

  /// The [commonMark] extension set is close to compliance with [CommonMark].
  ///
  /// [CommonMark]: http://commonmark.org/
  static final ExtensionSet commonMark = ExtensionSet(
    List<BlockSyntax>.unmodifiable(<BlockSyntax>[const FencedCodeBlockSyntax()]),
    List<InlineSyntax>.unmodifiable(<InlineSyntax>[InlineHtmlSyntax()]),
  );

  /// The [gitHubWeb] extension set renders Markdown similarly to GitHub.
  ///
  /// This is different from the [gitHubFlavored] extension set in that GitHub
  /// actually renders HTML different from straight [GitHub flavored Markdown].
  ///
  /// (The only difference currently is that [gitHubWeb] renders headers with
  /// linkable IDs.)
  ///
  /// [GitHub flavored Markdown]: https://github.github.com/gfm/
  static final ExtensionSet gitHubWeb = ExtensionSet(
     List<BlockSyntax>.unmodifiable(<BlockSyntax>[
       const FencedCodeBlockSyntax(),
       const HeaderWithIdSyntax(),
       const SetextHeaderWithIdSyntax(),
       const TableSyntax()
     ]),
     List<InlineSyntax>.unmodifiable(<InlineSyntax>[
       InlineHtmlSyntax(),
       StrikethroughSyntax(),
       EmojiSyntax(),
       AutolinkExtensionSyntax(),
     ])
  );

  /// The [gitHubFlavored] extension set is close to compliance with the [GitHub
  /// flavored Markdown spec].
  ///
  /// [GitHub flavored Markdown]: https://github.github.com/gfm/
  static final ExtensionSet gitHubFlavored = ExtensionSet(
    List<BlockSyntax>.unmodifiable(<BlockSyntax>[
      const FencedCodeBlockSyntax(),
      const TableSyntax()
    ]),
    List<InlineSyntax>.unmodifiable(<InlineSyntax>[
      InlineHtmlSyntax(),
      StrikethroughSyntax(),
      AutolinkExtensionSyntax(),
    ])
  );

  final List<BlockSyntax> blockSyntaxes;
  final List<InlineSyntax> inlineSyntaxes;

  ExtensionSet(this.blockSyntaxes, this.inlineSyntaxes);
}
```

**Comment by domesticmouse** [comment]
I like the idea of converting these sets to immutable. Do you want to put together a PR?

/cc @kevmoo

**Comment by srawlins** [comment]
We can make it immutable, but this would be a breaking change; I wonder if there is a way to make the mutability deprecated... tricky.

But I agree, we can make the extension sets immutable.

**Comment by mjordan56** [comment]
@srawlins Hi Sam. I agree this is a breaking change but it's trying to help the consumer of the markdown package from shooting themselves in the foot so to speak. If anyone is using the markdown package and is modifying either the blockSyntaxes or inlineSyntaxes list of the pre-defined extension sets they probably don't realize they're modifying a static property. In the case I debugged in the flutter_markdown package, a custom inline syntax parser to handle Markdown task list tags was created. This new inline syntax parser, TaskListSyntax, is added by the MarkdownWidget to the Document constructor:

```
    final md.Document document = md.Document(
      extensionSet: widget.extensionSet ?? md.ExtensionSet.gitHubFlavored,
      inlineSyntaxes: (widget.extensionSet?.inlineSyntaxes ?? [])
        ..add(TaskListSyntax())
        ..map((syntax) => syntax),
      encodeHtml: false,
    );
```
There are several things about this code I don't agree with and have submitted issues for but let's just focus on the inlineSyntaxes parameter. If the flutter_markdown package consumer creates an instance of Markdown in their Flutter app, this code snippet will be called every time the Markdown widget has a dependency change or is updated. If one of the pre-defined static property extension sets was specified by the user, that extension set will be modified with a growing list of TaskListSyntax objects. This is not a good thing. I contend things like this should break so they get fixed as soon as possible. This problem is a fairly silent evil lying in wait to cause problems down the road.

If you disagree with the break bad code argument, one possible solution that I believe is not a breaking change would be to change the pre-defined static properties to factory constructors. I haven't thought through all of the ramifications of this type of change but I believe it would be compatible with existing code and consumers of the package would be getting an instance of the extension set to use and modify and not a static that for all intents and purposes is a global variable.

As an aside for future reference, with the availability of Dart extensions now, I'd really like to see pre-defined types like this defined in an enum with additional functionality added with an extension. I did this in the enhanced example code I added to flutter_markdown. Here's the ExtensionSet wrapper I created:

```
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:markdown/markdown.dart' as md;

enum MarkdownExtensionSet { none, commonMark, githubFlavored, githubWeb }

extension MarkdownExtensionSetExtension on MarkdownExtensionSet {
  String get name => describeEnum(this);

  String get displayTitle => () {
        switch (this) {
          case MarkdownExtensionSet.none:
            return 'None';
          case MarkdownExtensionSet.commonMark:
            return 'Common Mark';
          case MarkdownExtensionSet.githubFlavored:
            return 'GitHub Flavored';
          case MarkdownExtensionSet.githubWeb:
            return 'GitHub Web';
        }
      }();

  md.ExtensionSet get value => () {
        switch (this) {
          case MarkdownExtensionSet.none:
            return md.ExtensionSet.none;
          case MarkdownExtensionSet.commonMark:
            return md.ExtensionSet.commonMark;
          case MarkdownExtensionSet.githubFlavored:
            return md.ExtensionSet.gitHubFlavored;
          case MarkdownExtensionSet.githubWeb:
            return md.ExtensionSet.gitHubWeb;
        }
      }();
}
```
The extension code could also include a copyOf method and other behaviors. ;)

**Comment by srawlins** [comment]
Fixed by #304
